internal/hcl/graph.go

```diff
@@ -385,16 +373,49 @@ func (g *Graph) AsJSON() ([]byte, error) {
 
 func (g *Graph) Walk() {
 	v := NewGraphVisitor(g.logger, g.vertexMutex)
+	TopologicalWalk(g.dag, v.Visit)
+}
 
-	flowCallback := func(d *dag.DAG, id string, parentResults []dag.FlowResult) (interface{}, error) {
-		vertex, _ := d.GetVertex(id)
-
-		v.Visit(id, vertex)
+func TopologicalWalk(graph *dag.DAG, visitor func(id string, vertex Vertex)) {
```

Comment:
I think the only thing that might be missing from this is a cycles check. This is built into the DescendantsFlow function. However, the last time I looked at the benchmarks the cycles check took a ton of mem to implement, so it might be best to rip it out, as you've done here. Just noting here for the sake of audit at a later date.